[cbj252] iP

cbj252 opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [x]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Description lifted from [CS2103T Website](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

cbj252 committed:
Duke Level-0

cbj252 committed:
Duke Level-1

cbj252 committed:
Duke Level-2

cbj252 committed:
Duke Level-3

cbj252 committed:
Duke Level-4

cbj252 committed:
Duke A-Testing

cbj252 committed:
Duke A-Exceptions

cbj252 committed:
Duke Level-6

cbj252 committed:
Duke Level-7

cbj252 committed:
Merge branch 'branch-Level-7'

cbj252 committed:
Made changes to do Duke-Level-8

cbj252 committed:
Merge branch 'branch-Level-8'

cbj252 committed:
Made new classes to solve Duke A-MoreOOP

cbj252 committed:
Put all files into duke package

cbj252 committed:
Update gitignore to ignore data files

cbj252 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

cbj252 committed:
Add gradle support

cbj252 committed:
Add Junit Tests

cbj252 committed:
Add Javadocs

cbj252 committed:
Add changes to do A-CodingStandard

cbj252 committed:
Add find command via regex checking

cbj252 committed:
Merge branch 'branch-A-CodingStandard'

cbj252 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java

cbj252 committed:
Remove need for deadline and event tasks to have a deadline, from or to requirement

cbj252 committed:
Make style changes after setting up automatic checkstyle

cbj252 committed:
Merge branch 'branch-A-Checkstyle'

cbj252 committed:
Create Java GUI without integration with old code

cbj252 committed:
Integrate old code with GUI

cbj252 committed:
Merge branch 'branch-Level-10'

cbj252 committed:
Fix bye and welcome command

dylkaw commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# duke.Duke project template
```
Is this a typo? I think Duke might have changed in some places to duke.Duke after refactoring.

dylkaw commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Completes the task with index taskNumber.
```
I like your javadocs! They are very clear and easy to understand. I like that you include any issues with the methods too.

dylkaw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,163 @@
+     * @return the task.
+     * @throws DukeException if incomplete task or not a task.
+     */
+    public static Task newTask(String response) throws DukeException {
```
Minor coding standard issue but should this be a verb?

Angelyxx replied:
Consider using helper methods at commands for each task type to improve code readability.

dylkaw reviewed (commented):
Good overall, just some very minor coding standard issues.

Angelyxx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,151 @@
+     * @return the task represented by the line from the data file.
+     * @throws DukeException if the data file's input does not match writing.
+     */
+    private Task parseOneLine(String oneTask) throws DukeException {
```
Consider using helper methods to improve the readability of the code.

Angelyxx reviewed (commented):
LGTM, minor improvements can be made to readability

cbj252 committed:
Complete A-Assertions by adding assertions throughout the code

Assertions were created immediately after the mark and delete
commands to ensure they were indeed marked as completed or
deleted respectively.

As for the find command, the current implementation relies on the
command being exactly four letters "find", followed by the keyword.
An assertion is placed in case this assumption becomes false later.

cbj252 committed:
Refactor parseOneLine & taskCreation

The two functions were previously responsible for parsing all
three types of tasks, leading to very long functions and arrowhead
style-coding. Helper functions have been added to mitigate this.

Some functions are also renamed, like newTask to createNewTask
to better fit the verb requirements.

cbj252 committed:
Merge pull request #1 from cbj252/branch-A-CodeQuality

Refactor parseOneLine & taskCreation

cbj252 committed:
Merge branch 'master' into branch-A-Assertions

cbj252 committed:
Merge pull request #2 from cbj252/branch-A-Assertions

Complete A-Assertions by adding assertions throughout the code

cbj252 committed:
Add extension of help command

cbj252 committed:
Change starting command to mention help command

cbj252 committed:
Update tests

cbj252 committed:
Bug fixes

cbj252 committed:
Add Ui image

cbj252 committed:
Rename chatbot from iP to botM

cbj252 committed:
Add user guide

cbj252 committed:
Fix README.md unwanted text

cbj252 committed:
Update README.md

cbj252 committed:
Fix Parser findCommand Javadoc

cbj252 committed:
Convert from Java 19 to Java 11

To fit course requirements

cbj252 committed:
Make find command print out messages with new lines

cbj252 committed:
Make bye command close the GUI

cbj252 committed:
Fix error messages not working
